File: 프로그래머스_코틀린/level2/프로세스.kt
package 프로그래머스_코틀린.level2

import kotlin.collections.ArrayDeque

class Solution56 {
    fun solution(priorities: IntArray, location: Int): Int {
        var answer = 0
        var queue = ArrayDeque<Pair<Int, Int>>()
        for ((idx, p) in priorities.withIndex()){
            queue.add(Pair(idx, p))
        }

        while (queue.size > 0){
            var firstEle = queue.removeFirst()
            if (queue.any { it.second > firstEle.second }){
                queue.add(firstEle)
            } else{
                answer ++
                if (location == firstEle.first) break
            }
        }
        return answer
    }
}

File: 프로그래머스/2/42587. 프로세스/프로세스.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <deque>
#include <iostream>

using namespace std;


int solution(vector<int> priorities, int location) {
    deque<pair<int, int>> dq;
    
    for (int i = 0; i < priorities.size(); i++){
        dq.push_back({priorities[i], i});
    }
    pair<int, int> max = *max_element(dq.begin(), dq.end());
    
    int idx = 1;
    while (true){
        pair<int, int> curr = dq[0];
        dq.pop_front();
        
        if (curr.first == max.first){
            if (curr.second == location){
                return idx;
            }
            idx ++;
            max = *max_element(dq.begin(), dq.end());
        }
        else{
            dq.push_back(curr);
        }
    }
}